src/emmy/expression/compile.cljc

```diff
@@ -296,61 +429,78 @@
+        ;; which are appended to `buffer`. If not, a return statement is generated for
+        ;; the expression at `z`. Nothing is returned.
+        do-body (fn [z]
+                  (if (and (z/branch? z) (= (z/node (z/next z)) `doto))
+                    (let [z (z/next (z/next z))
+                          var (z/node z)]
+                      (when-not (symbol? var)
+                        (throw (ex-info "Expecting a symbol (referring to a primitive array)"
+                                        {:unexpected var})))
+                      (doseq [[aset ix value] (z/rights z)]
+                        (when-not (= aset 'aset-double)
+                          (throw (ex-info "Expecting an aset-double statement"
+                                          {:unexpected aset})))
+                        (swap! buffer conj (str "  " var "[" ix "] = " (render/->JavaScript value) ";")))
+                      (z/next z))
+                    (swap! buffer conj
+                           (str "  return " (render/->JavaScript (z/node z)) ";"))))
+        ;; do-let accepts a zipper location. If the location is a `(let)` form,
+        ;; the bindings are unrolled into the corresponding sequence of assignment
+        ;; statements which are then appended to `buffer`. If the RHS of a binding
+        ;; pair is of the form `(aget var index)` this will be converted to an
+        ;; array reference; otherwise the usual Javascript renderer is used for the
+        ;; RHS of the assignment. The zipper position immediately following the let
+        ;; form is returned, facilitating this function's use in a parsing chain.
+        do-let (fn [z]
+                 (if (and (z/branch? z) (= (z/node (z/next z)) `let))
+                   (let [z (z/next (z/next z))]
+                     (doseq [[var value] (partition 2 (z/node z))]
+                       (if (and (seq? value) (= (first value) `aget))
+                         (swap! buffer conj (str "  const " var " = " (nth value 1) "[" (nth value 2) "];"))
+                         (swap! buffer conj (str "  const " var " = " (render/->JavaScript value) ";"))))
+                     (z/next z))
+                   z))
+        ]
+    (-> body z/seq-zip do-let do-body)
+    (conj argv (s/join "\n" @buffer))))
 
 (defn- compile-native
-  "Returns a natively-evaluated function that implements `body`, given:
-
-  - `argv`: a vector of symbols to serve as the function's arguments.
-    The vector may be nested for sequence destructuring.
-
-  - `code`: a map containing `bindings` and `body` keys"
-  [argv code]
-  #?(:clj (eval (compile->clj argv code))
-     :cljs (comp js->clj (apply js/Function (compile->js argv code)))))
+  "Dispatches the `code` object to the compiler corresponding to the
+  current runtime environment."
+  [code]
+  #?(:clj (eval (compile->clj code))
+     :cljs (comp js->clj (apply js/Function (compile->js code)))))
```

Comment:
do we still need this `js->clj` in every case? Not in the primitive calling convention case, right?